File: ml-benchmark/synduce/constraints/ensures/nested_minmax_noe.f
Inductive List = elt Int | cons {Int, List};
Inductive NList = line List | ncons {List, NList};
Inductive CNList = sglt List | cat {CNList, CNList};

cton = fix (
  \f: CNList -> NList.
  let dec = fix (
    \g: CNList -> CNList -> NList. \l: CNList. \c: CNList.
    match c with
      sglt x -> ncons {x, f l}
    | cat {x, y} -> g (cat {y, l}) x
    end
  ) in \c: CNList.
  match c with
    sglt x -> line x
  | cat {x, y} -> dec y x
  end
);


max = \a: Int. \b: Int. if < a b then b else a;
min = \a: Int. \b: Int. if < a b then a else b;

range = fix (
  \f: List -> {Int, Int}. \xs: List.
  match xs with
    elt w -> {w, w}
  | cons {h, t} ->
    let res = f t in
      {min h res.1, max h res.2}
  end
);

spec = fix (
  \f: NList -> {Int, Int}. \xs: NList.
  match xs with
    line a -> range a
  | ncons {h, t} ->
    let rh = range h in
      let res = f t in
        {min rh.1 res.1, max rh.2 res.2}
  end
);

target = fix (
  \f: CNList -> Compress CNList. \c: CNList.
  match c with
    sglt x ->
    let info = range x in /*This invocation is provided in Synduce's template*/
      c
  | cat {l, r} -> cat {f l, f r}
  end
);

main = \c: CNList. spec (cton (target c));

File: ml-benchmark/synduce/constraints/ensures/nested_minmax_noe.ml
type list =
  | Elt of int
  | Cons of int * list

type nlist =
  | Line of list
  | Ncons of list * nlist

type cnlist =
  | Sglt of list
  | Cat of cnlist * cnlist

let cton =
  let rec f c =
    let dec =
      let rec g l c2 =
        match c2 with
        | Sglt x -> Ncons (x, f l)
        | Cat (x, y) -> g (Cat (y, l)) x
      in
      g
    in
    match c with
    | Sglt x -> Line x
    | Cat (x, y) -> dec y x
  in
  f

let max a b = if a < b then b else a
let min a b = if a < b then a else b

let range =
  let rec f xs =
    match xs with
    | Elt w -> (w, w)
    | Cons (h, t) ->
        let res = f t in
        match res with
        | (a, b) -> (min h a, max h b)
  in
  f

let spec =
  let rec f xs =
    match xs with
    | Line a -> range a
    | Ncons (h, t) ->
        let rh = range h in
        let res = f t in
        match rh with
        | (rh1, rh2) ->
            match res with
            | (r1, r2) -> (min rh1 r1, max rh2 r2)
  in
  f

let target =
  let rec f c =
    match c with
    | Sglt x ->
        let info = range x in
        c
    | Cat (l, r) -> Cat (f l, f r)
  in
  f

let program c = spec (cton (target c))